### DISCUSSION TITLE
Middleware that sets response header based on request header

### DISCUSSION THREAD

**Discussion by dspicher** [tokio-rs/axum#1131] [Q&A] [answered]
I would like to implement a middleware that sets a response header to the identical value as the one received in the request.

Is there an easier way to achieve this than the one discussed under the scope of general state sharing [here](https://github.com/tokio-rs/axum/discussions/236#discussioncomment-1218395)?

**Comment by davidpdrsn** [comment, marked as answer]
You can use https://docs.rs/tower-http/latest/tower_http/propagate_header/index.html.

Otherwise you can do this:

```rust
async fn propagate_header<B>(req: Request<B>, next: Next<B>) -> Response {
    let header = req.headers().get("...").unwrap();
    let mut res = next.run(req).await;
    res.headers_mut().insert("key", header);
    res
}

Router::new().route(...).middleware(axum::middleware::from_fn(propagate_header));
```

**Reply by dspicher** [reply]
I knew my searching skills were failing me :see_no_evil: thank you so much for the quick answer! :heart:
